test/main_test.cpp

```diff
@@ -0,0 +1,218 @@
+        u2000.propertyName = 1000 + ( \
+        u1000.propertyName = 1000 + propertyId++)));
+      U(id) U(x) U(y) U(health) U(max_health) U(shield) U(max_shield)
+      U(energy) U(maxCD) U(groundCD) U(airCD) U(flags) U(type)
+      U(armor) U(shieldArmor) U(size) U(pixel_x) U(pixel_y)
+      U(pixel_size_x) U(pixel_size_y) U(groundATK) U(airATK)
+      U(groundDmgType) U(airDmgType) U(groundRange) U(airRange)
+      U(velocityX) U(velocityY)
+      U(playerId) U(resources) U(buildTechUpgradeType)
+      U(remainingBuildTrainTime) U(remainingUpgradeResearchTime) U(spellCD)
+      U(associatedUnit) U(associatedCount)
+      u5000.visible = u1000.visible = true;
+      u6000.visible = u2000.visible = false;
+      u1000.orders = {{ 10, 11, 12, 13, 14 }, { -10, -11, -12, -13, -14 }};
+      u2000.orders = {{ 20, 21, 22, 23, 24 }, { -20, -21, -22, -23, -24 }};
+      u5000.orders = {{ 50, 51, 52, 53, 54 }, { -50, -51, -52, -53, -54 }};
+      u6000.orders = {{ 60, 61, 62, 63, 64 }, { -60, -61, -62, -63, -64 }};
+      u1000.command = { 101, 102, 103, 104, 105, 106 };
+      u2000.command = { 201, 202, 203, 204, 205, 206 };
+      u5000.command = { 501, 502, 503, 504, 505, 506 };
+      u6000.command = { 601, 602, 603, 604, 605, 606 };
+  
+      // Our Frame is complete.
+      // Let's see how it handles serialization!
+      std::stringstream buffer;
+      buffer << frameBefore;
+      buffer >> frameAfter;
+
+      // Convert unsorted maps to sorted vectors for comparison
+      auto sortedResourcesBefore = sortMap(frameBefore.resources);
+      auto sortedResourcesAfter = sortMap(frameAfter.resources);
+      auto sortedActionsBefore = sortMap(frameBefore.actions);
+      auto sortedActionsAfter = sortMap(frameAfter.actions);
+      auto sortedUnitsBefore = sortMap(frameBefore.units);
+      auto sortedUnitsAfter = sortMap(frameAfter.units);
+
+      // Lest tries to fancy-print EXPECT arguments.
+      // But it but doesn't know how to print these complex objects.
+      // So we do the comparison here and feed Lest the results.
+      auto matchingResources = sortedResourcesBefore == sortedResourcesAfter;
+      auto matchingActions = sortedActionsBefore == sortedActionsAfter;
+      auto matchingUnits = sortedUnitsBefore == sortedUnitsAfter;
+
+      EXPECT(frameBefore.width == frameAfter.width);
+      EXPECT(frameBefore.height == frameAfter.height);
+      EXPECT(frameBefore.reward == frameAfter.reward);
+      EXPECT(frameBefore.is_terminal == frameAfter.is_terminal);
+      EXPECT(frameBefore.creep_map == frameAfter.creep_map);
+      EXPECT(frameBefore.bullets == frameAfter.bullets);
+      EXPECT(matchingResources);
+      EXPECT(matchingActions);
+      EXPECT(matchingUnits);
+    }
+  },
+  
+  lest_CASE("A TorchCraft FrameDiff is invariant through serialization") {
+    SETUP("Create & serialize a FrameDiff") {
+      torchcraft::replayer::FrameDiff diffBefore, diffAfter;
+      diffBefore.reward = 1.1;
```

Comment:
Reward is actually an integer. @syhw @ebetica that's the case for the C++ structs as well -- intended? Can we maybe just kill this field?

Reply:
I think both reward and is_terminal are left over from the E2D2 days, we can kill them in a future diff.